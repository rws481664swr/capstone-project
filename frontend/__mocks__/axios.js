import {BASE_URL} from "../src/config";

axios.get.mockImplementation((url) => {
    let result = url.slice(BASE_URL.length);
    if(result[result.length-1] === '/') result = result.slice(0, result.length-1);
    switch (result){
        case '/courses':
            return Promise.resolve({data: coursesList})
        case '/users':
            return Promise.resolve({data:usersList})
        case '/auth/token':
            return Promise.resolve({data:'token'})
    }
    return Promise.resolve({data: []})

})
const axios = {
    get: jest.fn().mockResolvedValue({
        data: {}
    }),
    post: jest.fn().mockResolvedValue({
        data: {}
    }),
    put: jest.fn().mockResolvedValue({
        data: {}
    }),
    delete: jest.fn().mockResolvedValue({
        data: {}
    })
}
export default axios