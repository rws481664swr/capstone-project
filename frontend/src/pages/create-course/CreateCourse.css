.unauthorized.CourseFormFlash{
    display: flex;
    justify-content: center;
}