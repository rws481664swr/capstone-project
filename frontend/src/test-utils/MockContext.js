import GlobalContext from "../state/contexts/GlobalContext";
import {MOCK_TOKEN} from "./mock-backend";
import {MemoryRouter} from "react-router-dom";

const MockContext = ({children, isAdmin = false, ...mockValue}) => {
    return (
        <MemoryRouter>
            <GlobalContext.Provider value={{
                token: MOCK_TOKEN,
                _id: 'id',
                username: 'username',
                role: isAdmin ? "ADMIN" : "STUDENT",
                loggedIn: true,
                timestamp: new Date(),
                ...mockValue
            }}>
                {children}
            </GlobalContext.Provider>
        </MemoryRouter>
    )
}

export default MockContext