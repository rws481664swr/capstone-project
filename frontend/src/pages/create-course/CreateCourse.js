import './CreateCourse.css'
import LabeledInput from "../../components/General/LabeledInput/LabeledInput";
import Button from "../../components/General/Button/GenericButton/Button";
import useCreateCourse from "./useCreateCourse";
import {useGlobalContext} from "../../state/contexts/GlobalContext";


const CreateCourse = () => {
    const {role} = useGlobalContext()
    const [{form, onChange}, {toRender, setCss, error}, submit] = useCreateCourse()
    const isAdmin = role === "ADMIN"
    const isTeacher = role === "TEACHER"
    if (!(isAdmin || isTeacher)) return (
        <div id={'unauthorized-courses-new'}
             className={'unauthorized sr-container'}>
            <h1>Unauthorized</h1>
        </div>
    )
    return (
        <div className={'sr-container'}>
            <h1>Create A New Course</h1>
            <form onSubmit={submit}>

                {error && <div className={'CourseFormFlash'}>{error}</div>}
                <LabeledInput
                    name={'courseName'}
                    value={form.courseName}
                    onChange={onChange}
                    label={'Course Name'}
                    id={'input-courseName'}
                    data-testid={'input-create-courseName'}
                    flash={toRender.courseName}
                />
                <LabeledInput
                    name={'courseNumber'}
                    value={form.courseNumber}
                    onChange={onChange}
                    label={'Course Number'}
                    id={'input-courseNumber'}
                    data-testid={'input-create-courseNumber'}
                    type={'number'}
                    flash={toRender.courseNumber}
                />
                <LabeledInput
                    name={'subject'}
                    value={form.subject}
                    onChange={onChange}
                    label={'Subject'}
                    id={'input-subject'}
                    data-testid={'input-create-subject'}
                    flash={toRender.subject}
                />
                <LabeledInput
                    name={'startDate'}
                    value={form.startDate}
                    onChange={onChange}
                    label={'Start Date'}
                    id={'input-startDate'}
                    data-testid={'input-create-startDate'}
                    type={'date'}
                    flash={toRender.startDate}
                />
                <LabeledInput
                    name={'endDate'}
                    value={form.endDate}
                    onChange={onChange}
                    label={'End Date'}
                    id={'input-endDate'}
                    data-testid={'input-create-endDate'}
                    type={'date'}
                    flash={toRender.endDate}
                />

                <Button id={'submit-course'} type={'submit'}>
                    Create Course
                </Button>
            </form>
        </div>
    )
}

export default CreateCourse