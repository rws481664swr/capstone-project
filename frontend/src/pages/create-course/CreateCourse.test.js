import {act, fireEvent, render} from "@testing-library/react";
import MockContext from "../../test-utils/MockContext";


import CreateCourse from "./CreateCourse";
import axios from "axios";
import reactRouterDom from "react-router-dom";

jest.mock('axios')
const mockNavigate = jest.fn(() => ({id: null}))
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(
        () => mockNavigate)
}))
describe('CreateCourse', () => {
    it('should render CreateCourse', () => {
        const {getByTestId, getByLabelText, getByText} = render(
            <MockContext role={"TEACHER"}>
                <CreateCourse/>
            </MockContext>)
        expect(getByLabelText('Course Name')).toBeInTheDocument()
        expect(getByLabelText('Subject')).toBeInTheDocument()
        expect(getByLabelText('Start Date')).toBeInTheDocument()
        expect(getByLabelText('End Date')).toBeInTheDocument()

    })
    it('should match snapshot', () => {
        const {asFragment} = render(
            <MockContext role={"TEACHER"}>
                <CreateCourse/>
            </MockContext>)
        expect(asFragment()).toMatchSnapshot()
    })
    it('should render error message if not a teacher or admin', () => {
        const {getByText} = render(
            <MockContext role={"STUDENT"}>
                <CreateCourse/>
            </MockContext>)
        expect(getByText('Unauthorized')).toBeInTheDocument()
    })
    it('should create a course', async () => {
        const {getByTestId, getByLabelText, getByText} = render(
            <MockContext role={"TEACHER"}>
                <CreateCourse/>
            </MockContext>)
        const nav = jest.fn()
        const mockNav=() => nav
        reactRouterDom.useNavigate.mockImplementationOnce(mockNav)
        axios.post.mockImplementationOnce(async (url,body,config) => {
            expect(url).toInclude('courses')
            expect(body).toHaveProperty('courseName')
            expect(body).toHaveProperty('subject')
            expect(body).toHaveProperty('startDate')
            expect(body).toHaveProperty('endDate')

            return {data: {message: 'Course Created'}}
        }   )
        fireEvent.change(getByLabelText('Course Name'), {target: {value: 'test'}})
        fireEvent.change(getByLabelText('Subject'), {target: {value: 'test'}})
        fireEvent.change(getByLabelText('Start Date'), {target: {value: '2021-01-01'}})
        fireEvent.change(getByLabelText('End Date'), {target: {value: '2021-01-01'}})
        fireEvent.click(getByText('Create Course'))
        await act(() => fireEvent.click(getByText('Create Course')))
    })

})