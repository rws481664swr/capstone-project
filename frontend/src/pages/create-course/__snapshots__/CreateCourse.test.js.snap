// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCourse should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sr-container"
  >
    <h1>
      Create A New Course
    </h1>
    <form>
      <div
        class="LabeledInput-div "
      >
        <label
          class="LabeledInput-input-label "
          for="input-courseName"
        >
          Course Name
        </label>
        <div />
        <input
          class="LabeledInput-input-text "
          data-testid="input-create-courseName"
          id="input-courseName"
          name="courseName"
          type="text"
          value=""
        />
      </div>
      <div
        class="LabeledInput-div "
      >
        <label
          class="LabeledInput-input-label "
          for="input-courseNumber"
        >
          Course Number
        </label>
        <div />
        <input
          class="LabeledInput-input-text "
          data-testid="input-create-courseNumber"
          id="input-courseNumber"
          name="courseNumber"
          type="number"
          value="0"
        />
      </div>
      <div
        class="LabeledInput-div "
      >
        <label
          class="LabeledInput-input-label "
          for="input-subject"
        >
          Subject
        </label>
        <div />
        <input
          class="LabeledInput-input-text "
          data-testid="input-create-subject"
          id="input-subject"
          name="subject"
          type="text"
          value=""
        />
      </div>
      <div
        class="LabeledInput-div "
      >
        <label
          class="LabeledInput-input-label "
          for="input-startDate"
        >
          Start Date
        </label>
        <div />
        <input
          class="LabeledInput-input-text "
          data-testid="input-create-startDate"
          id="input-startDate"
          name="startDate"
          type="date"
          value=""
        />
      </div>
      <div
        class="LabeledInput-div "
      >
        <label
          class="LabeledInput-input-label "
          for="input-endDate"
        >
          End Date
        </label>
        <div />
        <input
          class="LabeledInput-input-text "
          data-testid="input-create-endDate"
          id="input-endDate"
          name="endDate"
          type="date"
          value=""
        />
      </div>
      <button
        class="undefined Button"
        id="submit-course"
        type="submit"
      >
        Create Course
      </button>
    </form>
  </div>
</DocumentFragment>
`;
